tensorflow/core/lib/random/random_distributions.h

```diff
@@ -95,47 +138,53 @@ template <class Generator>
 class UniformDistribution<Generator, bfloat16> {
  public:
   // The number of elements that will be returned.
-  static constexpr int kResultElementCount = Generator::kResultElementCount;
+  // Set the number to be Eigen packet size of type at least,
+  // so computations can be vectorized using SIMD.
+  static constexpr int kVectorLength =
+      Eigen::internal::packet_traits<bfloat16>::size;
+  static constexpr int kResultElementCount =
+      std::max(kVectorLength, Generator::kResultElementCount);
   // Cost of generation of a single element (in cycles).
   static constexpr int kElementCost = 3;
   // Indicate that this distribution may take variable number of samples
   // during the runtime.
   static constexpr bool kVariableSamplesPerOutput = false;
   typedef Array<bfloat16, kResultElementCount> ResultType;
   typedef bfloat16 ResultElementType;
+  // Helper definition for the format function.
+  typedef bfloat16 (*FormatFunc)(uint16);
 
   PHILOX_DEVICE_INLINE
   ResultType operator()(Generator* gen) {
-    typename Generator::ResultType sample = (*gen)();
-    ResultType result;
-    for (int i = 0; i < kResultElementCount; ++i) {
-      result[i] = Uint16ToGfloat16(sample[i]);
-    }
-    return result;
+    return VectorizedFormat<UniformDistribution<Generator, bfloat16>,
+                            Generator>(gen, InternalUint16ToBfloat16);
   }
 };
 
 template <class Generator>
 class UniformDistribution<Generator, float> {
  public:
   // The number of elements that will be returned.
-  static constexpr int kResultElementCount = Generator::kResultElementCount;
+  // Set the number to be Eigen packet size of type at least,
+  // so computations can be vectorized using SIMD.
+  static constexpr int kVectorLength =
+      Eigen::internal::packet_traits<float>::size;
+  static constexpr int kResultElementCount =
+      std::max(kVectorLength, Generator::kResultElementCount);
   // Cost of generation of a single element (in cycles).
   static constexpr int kElementCost = 3;
   // Indicate that this distribution may take variable number of samples
   // during the runtime.
   static constexpr bool kVariableSamplesPerOutput = false;
   typedef Array<float, kResultElementCount> ResultType;
   typedef float ResultElementType;
+  // Helper definition for the format function.
+  typedef float (*FormatFunc)(uint32);
 
   PHILOX_DEVICE_INLINE
   ResultType operator()(Generator* gen) {
-    typename Generator::ResultType sample = (*gen)();
-    ResultType result;
-    for (int i = 0; i < kResultElementCount; ++i) {
-      result[i] = Uint32ToFloat(sample[i]);
-    }
-    return result;
+    return VectorizedFormat<UniformDistribution<Generator, float>, Generator>(
```

Comment:
@Zantares here is the internal error , can you please check 

`random_distributions.h:185:12: error: reference to __host__ function 'VectorizedFormat<tensorflow::random::UniformDistribution<tensorflow::random::PhiloxRandom, float>, tensorflow::random::PhiloxRandom>' in __host__ __device__ function
    return VectorizedFormat<UniformDistribution<Generator, float>, Generator>(`